docs: Plans (#2270)
block/goose: documentation/docs/guides/creating-plans.md

---
sidebar_position: 17

> title: Creating Plans with the /plan prompt

Reviewer: ```suggestion
title: Creating Plans Before Working
```

sidebar_label: Creating Plans
---

Starting a project without a clear plan is like building a house without a blueprint. It can lead to:

* Confusion about what to do
* Wasted time and effort
* Projects that grow too big

A good plan keeps everyone on track and helps measure progress. That's why the Goose CLI includes the `/plan` prompt completion command to help break down your projects into clear, manageable steps.
 

> :::tip Plans in the Goose desktop

Reviewer 2: ```suggestion
:::tip Plans in the Goose Desktop
```

Reviewer 2: for consistency in our docs..let's write Goose Desktop -- capitalized..but this is a good read thus far.

> The Goose desktop doesn't have a `plan` keyword. If you want the desktop to create a plan for you, you need to use a prompt like:

Reviewer 2: what if we used tabs here..one for Goose Desktop and one for Goose CLI ..similar to how we do with our other docs

Reviewer: had the same thought

```
"Hey Goose, can you create a plan to convert my CLI project into a locally hosted web page that gives me input fields for each CLI command I can run? Please don't start the actual work"
```
Unless you ask Goose to "create a plan", it might just start into the project work. 
:::

The Goose CLI's plan mode differs from desktop planning in a key way: it's interactive, asking clarifying questions to understand your project before creating a plan. In contrast, the desktop version creates plans immediately based on your description alone, making assumptions that might miss critical project details.

## Set your planner provider and model

> The Goose CLI plan mode uses two configuration values:

Reviewer: ```suggestion
In some workflows, it can be helpful to use one LLM for planning and a different one for execution. For example, GPT-4.1 tends to excel at strategic planning and breaking down complex tasks into clear, logical steps. On the other hand, Claude Sonnet 3.5 is particularly strong at writing clean, efficient code and following instructions precisely. By using GPT-4.1 to plan and Claude to execute, you can play to the strengths of both models and get better results overall.

The Goose CLI plan mode uses two configuration values:
```

want to provide more info on why someone would want to set these.

- `GOOSE_PLANNER_PROVIDER`: Which provider to use for planning
- `GOOSE_PLANNER_MODEL`: Which model to use for planning

For example, you might configure:
```bash
export GOOSE_PLANNER_PROVIDER=databricks

> export GOOSE_PLANNER_MODEL=goose-claude-3-7-sonnet

Reviewer: use an external model name. this one is specific to Block

```
If these aren't set, Goose will use your default provider and model settings. You might want to set different planning models if you find certain models are better at breaking down tasks into clear steps. However, your default model configuration is usually sufficient.

> To verify that the planner provider is set input the following terminal command:

Reviewer: ```suggestion
To verify that the planner provider is set, input the following terminal command:
```

```bash
~ goose info -v
```

In this example, the `info` command returns the current configuration and the path to the configuration file.  

```bash
Goose Version:
  Version:          1.0.18

Goose Locations:
  Config file:      /Users/alincoln/.config/goose/config.yaml
  Sessions dir:     /Users/alincoln/.local/share/goose/sessions
  Logs dir:         /Users/alincoln/.local/state/goose/logs

Goose Configuration:

>   GOOSE_PROVIDER: databricks

Reviewer: ```suggestion
  GOOSE_PROVIDER: anthropic
```

>   DATABRICKS_HOST: https://some-domain.databricks.com/

Reviewer: ```suggestion

```

>   DATABRICKS_HOST: https://block-lakehouse-production.cloud.databricks.com/

Reviewer: please remove

>   GOOSE_MODEL: databricks-meta-llama-3-3-70b-instruct

Reviewer: ```suggestion
  GOOSE_MODEL: claude-3.5-sonnet
```

>   GOOSE_PLANNER_PROVIDER: databricks

Reviewer: ```suggestion
  GOOSE_PLANNER_PROVIDER: openai
```

  GOOSE_MODE: smart_approve

>   GOOSE_PLANNER_MODEL: goose-claude-3-7-sonnet

Reviewer: ```suggestion
  GOOSE_PLANNER_MODEL: gpt-4.1
```

>   GOOSE_PLANNER_MODEL: goose-claude-3-7-sonnet

Reviewer: change to external model

```

>  If either `GOOSE_PLANNER_PROVIDER` or `GOOSE_PLANNER_MODEL` are not set, `GOOSE_PROVIDER` and `GOOSE_MODEL`are used to build your plan.  

Reviewer 2: This line reads a bit confusing to me.

## Describe your project

> While Goose can handle complex project descriptions, it works best with clear, concise ones. Focus on stating your project's purpose and desired outcomes – if these aren't clear, Goose will ask clarifying questions until it fully understands your goals. 

Reviewer: ```suggestion
While Goose can handle complex project descriptions, it works best with clear, concise ones. Focus on stating your project's purpose and desired outcomes. If these aren't clear, Goose will ask clarifying questions until it fully understands your goals. 
```

If Goose believes your project can be completed in many different ways and using a wide variety of components, it will ask you a clarifying question for each of these decision points. For example, if you start a plan like this:

```bash 
( O)> /plan

Entering plan mode. You can provide instructions to create a plan and then act on it. To exit early, type /endplan

( O)> Convert the CLI built by search_replace_routes.py into a web page
```
Goose parses your project description, consults with the LLM mode you've configured, and then if it needs more information, starts a round of clarifying questions.

## Clarifying questions

> Converting a Python CLI into a web site seems simple enough but Goose will have questions about things like styling, authentication, features, technology stack, and more. You might see questions like this:

Reviewer: ```suggestion
Converting a Python CLI into a website seems simple enough but Goose will have questions about things like styling, authentication, features, technology stack, and more. You might see questions like this:
```

```bash
1. Should the application support any keyboard shortcuts for common actions?
2. Would you like the application to remember user preferences (like case sensitivity setting) between sessions?
3. Should there be any form of notification when operations complete successfully?
4. How should the application handle very large text fields that might be difficult to display in the three-column layout?
5. Are there any specific CI/CD requirements for deployment to AWS?

```
You can answer the questions one at a time or you can batch your answers:

```bash
( O)> 1 no keyboard shortcuts. 2 do not remember preferences. 3 on success, open a dialog that says "success". 4 Truncate to 30 characters before the string to replace and then 30 characters after the string to replace.5 no CI/CD requirements 
```

:::tip

> When Goose requests a project artifact like source code during plan mode, you'll need to paste the content directly into the chat. Simply copying the file contents and prefixing it with a brief description like 'Here's the requested code:' is sufficient – providing just a file path won't work in plan mode.

Reviewer: ```suggestion
When Goose requests a project artifact like source code during plan mode, you'll need to paste the content directly into the chat. Simply copying the file contents and prefixing it with a brief description like 'Here's the requested code:' is sufficient. Note that providing just a file path won't work in plan mode.
```

---
sidebar_position: 17
title: Creating Plans Before Working
sidebar_label: Creating Plans
---

Starting a project without a clear plan is like building a house without a blueprint. It can lead to:

* Confusion about what to do
* Wasted time and effort
* Projects that grow too big

A good plan keeps everyone on track and helps measure progress. That's why the Goose CLI includes the `/plan` prompt completion command to help break down your projects into clear, manageable steps.
 
:::tip Plans in the Goose Desktop

> The Goose desktop doesn't have a `plan` keyword. If you want the desktop to create a plan for you, you need to use a prompt like:

Reviewer 2: ```suggestion
The Goose Desktop doesn't have a `plan` keyword. If you want Goose Desktop to create a plan for you, you need to use a prompt like:
```

```
"Hey Goose, can you create a plan to convert my CLI project into a locally hosted web page that gives me input fields for each CLI command I can run? Please don't start the actual work"
```
Unless you ask Goose to "create a plan", it might just start into the project work. 
:::

The Goose CLI's plan mode differs from desktop planning in a key way: it's interactive, asking clarifying questions to understand your project before creating a plan. In contrast, the desktop version creates plans immediately based on your description alone, making assumptions that might miss critical project details.

## Set your planner provider and model

> In some workflows, it can be helpful to use one LLM for planning and a different one for execution. For example, GPT-4.1 tends to excel at strategic planning and breaking down complex tasks into clear, logical steps. On the other hand, Claude Sonnet 3.5 is particularly strong at writing clean, efficient code and following instructions precisely. By using GPT-4.1 to plan and Claude to execute, you can play to the strengths of both models and get better results overall.

Reviewer 2: While I often suggest this to people..where I say oh this llm may be better for this particular situation..I wonder if this paragraph belongs in this doc. I feel we can create a best practices with LLMs doc or blog instead. Let's just jump into how to set the planner and provider model

Reviewer: The thought here is to let people know *why* they would even want/need to set different provider/models. I dont think that's obvious to without this paragraph, at least not to people who are unfamiliar with multi-modal workflows

Reviewer 2: oh okay. thank you @angiejones for the clarification. I see now that you're saying their will be two values GOOSE_PROVIDER for execution and GOOSE_PLANNER_PROVIDER for planning. And you it's best to choose an LLM that's good at planning for the provider variable..and use an LLM that's good at execution for GOOSE_PROVIDER

Reviewer 2: okay, so this is a total nitpick, but I'm neurodivergent..and wonder if other neurodivergent people might also be confused and miss the point like I did. I think I would follow if it was like 
```
## Set your planner provider and model
### Something about how to set them
Ensure the following variables are set 
GOOSE_PLANNER_PROVIDER and GOOSE_PLANNER_MODEL for planning
GOOSE_PROVIDER and GOOSE_MODEL for execution

###Something about why we're setting it different that the execution mode.
[your amazing paragraph explaining the strengths of different LLMs]

### then the check your config paragraph
```

just throwing it out there, but I trust your judgement

The Goose CLI plan mode uses two configuration values:

- `GOOSE_PLANNER_PROVIDER`: Which provider to use for planning
- `GOOSE_PLANNER_MODEL`: Which model to use for planning

> 

Reviewer 2: is it the case that the model needs to match the provider for example..can i mix open ai with claude-sonnet-3.5? or do i have to do anthropic with claude-sonnet-3.5 if the latter is the case..I think we should specify that

export GOOSE_PLANNER_PROVIDER=<my-chosen-provider>
export GOOSE_PLANNER_MODEL=<my-chosen-model>
```
If these aren't set, Goose will use your default provider and model settings. You might want to set different planning models if you find certain models are better at breaking down tasks into clear steps. However, your default model configuration is usually sufficient.

To verify that the planner provider is set, input the following terminal command:

```bash
~ goose info -v
```

In this example, the `info` command returns the current configuration and the path to the configuration file.  

```bash
Goose Version:
  Version:          1.0.18

Goose Locations:
  Config file:      /Users/alincoln/.config/goose/config.yaml
  Sessions dir:     /Users/alincoln/.local/share/goose/sessions
  Logs dir:         /Users/alincoln/.local/state/goose/logs

Goose Configuration:
  GOOSE_PROVIDER: anthropic
  GOOSE_MODEL: claude-3.5-sonnet
  GOOSE_PLANNER_PROVIDER: openai
  GOOSE_MODE: smart_approve
  GOOSE_PLANNER_MODEL: gpt-4.1
```

 If either `GOOSE_PLANNER_PROVIDER` or `GOOSE_PLANNER_MODEL` are not set, `GOOSE_PROVIDER` and `GOOSE_MODEL`are used to build your plan.  

## Describe your project

> While Goose can handle complex project descriptions, it works best with clear, concise ones. Focus on stating your project's purpose and desired outcomes. If these aren't clear, Goose will ask clarifying questions until it fully understands your goals. 

Reviewer 2: could the examples start really simple first like here's a simple plan and then do one with a more complex plan where it will ask clarifying questions

1. Create a service for interacting with Contentful
## Step 5: Implement local storage service for history
1. Create a service for managing search history in localStorage
## Step 6: Implement validation utilities
1. Create utility functions for input validation
## Step 7: Implement main pages
1. Create the Search page
2. Create the History page
## Step 8: Set up routing and main application
1. Create the main App component
## Step 9: Implement error handling and loading states
1. Create error boundary components
2. Add loading indicators for API operations
3. Implement error messages display

## Step 10: Set up deployment configuration
1. Create AWS deployment configuration
## Step 11: Testing and quality assurance
1. Write unit tests for key components
2. Implement integration tests for the main workflows
3. Perform manual testing of the application

This plan provides a comprehensive framework for converting your CLI script to a React web application with Material UI and Okta authentication. You'll need to adapt specific parts based on your exact requirements and the functionality of your original script.

```

## A simple construction plan example

> Goose can produce good plans for relatively simple projects such as the home construction example:

Reviewer 2: yeah maybe we can bring this example up as the first example..and then put the more complex examples near the bottom..so we can incrementally show /plan capabilities
